lean4-infoview/src/infoview/goalLocation.tsx

```diff
@@ -76,66 +76,125 @@ export interface Locations {
- * highlighted when hovered over if `alwaysHighlight` is `true` or `locs` and `loc` are both defined.
- * `setHoverState` is passed through to {@link DetectHoverSpan}. */
-export function SelectableLocation(props_: SelectableLocationProps): JSX.Element {
-    const { locs, loc, alwaysHighlight, setHoverState: setParentHoverState, ...props } = props_
-
-    const shouldHighlight: boolean = alwaysHighlight || (!!locs && !!loc)
+ * Logic for a `<span>` with a corresponding {@link GoalsLocation} that can be (un)selected using shift-click.
+ */
+export function useHighlightedLocation(settings: LocationHighlightSettings): HighlightedLocation {
+    const { ref, hoverSettings, modHoverSettings, selectionSettings } = settings
+
     const [hoverState, setHoverState] = React.useState<HoverState>('off')
-    let spanClassName: string = ''
-    if (shouldHighlight) {
-        spanClassName += 'highlightable '
-        if (hoverState !== 'off') spanClassName += 'highlight '
-        if (props.className) spanClassName += props.className
+
+    const locs = React.useContext(LocationsContext)
+    let className: string = ''
+    if (hoverSettings.highlightOnHover && hoverState !== 'off') {
+        className += 'highlight '
+    } else if (selectionSettings.highlightOnSelection && locs && locs.isSelected(selectionSettings.loc)) {
+        className += 'highlight-selected '
+    }
+    if (modHoverSettings.highlightOnModHover && hoverState === 'ctrlOver') {
+        className += 'underline '
+    }
+
+    const onPointerEvent = (b: boolean, e: React.PointerEvent<HTMLSpanElement>) => {
+        // It's more composable to let pointer events bubble up rather than to call `stopPropagation`,
+        // but we only want to handle hovers in the innermost component. So we record that the
+        // event was handled with a property.
+        // The `contains` check ensures that the node hovered over is a child in the DOM
+        // tree and not just a logical React child (see useLogicalDom and
+        // https://reactjs.org/docs/portals.html#event-bubbling-through-portals).
+        if (ref.current && e.target instanceof Node && ref.current.contains(e.target)) {
+            if ('_DetectHoverSpanSeen' in e) {
+                return
+            }
+            ;(e as any)._DetectHoverSpanSeen = {}
+            if (!b) {
+                setHoverState('off')
+            } else if (e.ctrlKey || e.metaKey) {
+                setHoverState('ctrlOver')
+            } else {
+                setHoverState('over')
+            }
+        }
+    }
+
+    const onPointerMove = (e: React.PointerEvent<HTMLSpanElement>) => {
+        if (e.ctrlKey || e.metaKey) {
+            setHoverState(st => (st === 'over' ? 'ctrlOver' : st))
+        } else {
+            setHoverState(st => (st === 'ctrlOver' ? 'over' : st))
+        }
+    }
+
+    const onKeyDown = (e: React.KeyboardEvent<HTMLSpanElement>) => {
```

Comment:
I'm not sure whether having a few global event handlers is really much better in practice (IIRC removing these handlers here doesn't make much of a difference). We don't prevent it here, but I think one argument for "local" event handlers is also that events don't need to bubble up the whole DOM when fired. 

In any case, this would be worth benchmarking if we run into performance issues again and want to squeeze out the last bit of performance - for now I prefer the simplicity of these local event handlers.

Reply:
Makes sense. Btw, I noticed a minor loss of functionality after this change: before, if you hovered over a span that has an associated definition and pressed ctrl, just doing the keypress would render the underline. Now, you have to move the cursor; otherwise the underline state will stay stale. This is obscure enough that it's fine to ignore imo, but maybe you have a quick fix.

Reply:
Good catch. Looks like the `div`-local `keyDown` and `keyUp` handlers here do not work at all because they are intended to trigger when the `div` is focused, which also explains why they used effects before :-(

I'll fix this by using effects that are conditional on the hover state again so that both performance and functionality are covered.